### DISCUSSION TITLE
form data extraction

### DISCUSSION THREAD

**Discussion by 88arvin** [jsvine/pdfplumber#950] [Ask for help with specific PDFs]
how to extract data (field name + field values) from filled form and save it into csv file?

**Comment by jsvine** [comment]
Hi @88arvin; have you tried the method described in the  "[Extracting form values](https://github.com/jsvine/pdfplumber#extracting-form-values)" section of the README.md? If so, can you describe what didn't work, and provide the PDF it didn't work for?

**Reply by 88arvin** [reply]
yess....getting this error `AttributeError: 'dict' object has no attribute 'resolve'`
[Form.pdf](https://github.com/jsvine/pdfplumber/files/12197650/Form.pdf)

**Reply by cmdlineluser** [reply]
I cannot replicate the error.
```python
pdf = pdfplumber.open('Downloads/Form.pdf')

fields = pdf.doc.catalog["AcroForm"].resolve()["Fields"]

form_data = {}

for field in fields:
    field_name = field.resolve()["T"]
    field_value = field.resolve()["V"]
    form_data[field_name] = field_value

print(form_data)
# {}
```
`Form.pdf` just has a please wait message in it.

**Reply by 88arvin** [reply]
Are you opening `Form.pdf` in any browser? If so, then it will not open because it is an XFA-based PDF and will only open with Adobe Acrobat.

**Reply by cmdlineluser** [reply]
Ah, I see. I was using `Preview.app` - I don't have Adobe Acrobat.

**Reply by 88arvin** [reply]
Okkk...is PDFplumber able to extract data from this type of PDF form?

**Reply by cmdlineluser** [reply]
I'm not entiely sure.

From what I can find, it suggests not.

> pdfplumber doesn't have an interface for working with form data, but you can access it using pdfplumber's wrappers around pdfminer

https://pdfminersix.readthedocs.io/en/latest/howto/acro_forms.html

> Only AcroForm interactive forms are supported, XFA forms are not supported.

**Reply by jsvine** [reply]
Thanks, @88arvin and @cmdlineluser. I'm not very familiar with XFA forms. I'd be theoretically open to adding support for them, but it might take me a while to familiarize myself with this corner of the PDF world. If there's someone out there that knows the XFA spec well, I'd love to hear from you.

In the meantime, you might have some luck with this particular PDF exploring the object here:

```python
pdf = pdfplumber.open("Form.pdf")
print(pdf.doc.catalog["AcroForm"].resolve()["XFA"])
```

**Reply by 88arvin** [reply]
Why I am getting this error - `AttributeError: 'dict' object has no attribute 'resolve'`
@cmdlineluser - Are you able to access `Form_new.pdf` ?
[Form_new.pdf](https://github.com/jsvine/pdfplumber/files/12212149/Form_new.pdf)

**Reply by cmdlineluser** [reply]
Yes, I can access that PDF.

I'm not sure why you're getting that error, I get `KeyError: 'AcroForm'` 
```python
pdf = pdfplumber.open('Downloads/Form_new.pdf')

fields = pdf.doc.catalog["AcroForm"].resolve()["Fields"]

form_data = {}

for field in fields:
    field_name = field.resolve()["T"]
    field_value = field.resolve()["V"]
    form_data[field_name] = field_value
```

**Reply by 88arvin** [reply]
@cmdlineluser and @jsvine 
Take a look at this - https://github.com/AF-VCD/pdf-xfa-tools

**Reply by cmdlineluser** [reply]
Yes, it uses `pikepdf` which uses `qpdf` to do the parsing.

https://github.com/pikepdf/pikepdf

https://github.com/qpdf/qpdf

As for `pdfminer`, https://github.com/pdfminer/pdfminer.six/issues/497 suggests it should be possible and refers to: https://stackoverflow.com/questions/18587878/extract-xdp-or-xfa-from-pdf/33293026#33293026

But nothing seems to have come from it as of yet.
